node/core/pvf/src/execute/worker.rs

```diff
@@ -104,25 +122,47 @@ pub async fn start_work(
 					);
 					return Outcome::IoErr
 				},
-				Ok(response) => response,
+				Ok(response) => {
+					if let Response::Ok{duration, ..} = response {
+						if duration > execution_timeout {
+							// The job didn't complete within the timeout.
+							gum::warn!(
+								target: LOG_TARGET,
+								worker_pid = %pid,
+								"execute job took {}ms cpu time, exceeded execution timeout {}ms.",
+								duration.as_millis(),
+								execution_timeout.as_millis(),
+							);
+
+							// Return a timeout error.
+							return Outcome::HardTimeout;
+						}
+					}
+
+					response
+				},
 			}
 		},
-		_ = Delay::new(execution_timeout).fuse() => {
+		_ = Delay::new(timeout).fuse() => {
 			gum::warn!(
 				target: LOG_TARGET,
 				worker_pid = %pid,
 				validation_code_hash = ?artifact.id.code_hash,
 				"execution worker exceeded alloted time for execution",
 			);
-			return Outcome::HardTimeout;
+			// TODO: This case is not really a hard timeout as the timeout here in the host is
+			// lenient. Should fix this as part of
```

Comment:
I actually don't understand this comment. This looks like a "hard" timeout to me.

Reply:
It was maybe poorly explained or wrong choice of words, but my thinking here was that the hard timeout is on the child, whereas here on the host it is a lenient timeout. (The comment is specifically referring to the `HardTimeout` error we return further down.) The distinction matters to me because reaching the lenient timeout may indicate that the child stalled, which we may want to treat differently. At any rate, this is something I will revisit soon when I get to the linked issue.